File: FrontEnd (C#)/SoftProgPersistance/GestMaterial/Impl/EjemplarImpl.cs
using SoftProgDBManager;
using SoftProgModel.GestMaterial;
using SoftProgPersistance.GestMaterial.DAO;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SoftProgPersistance.GestMaterial.Impl
{
    public class EjemplarImpl : EjemplarDAO
    {
        private DbDataReader lector;
        public int eliminar(int idObjeto)
        {
            DbParameter[] parametros = new DbParameter[1];
            parametros[0] = DBManager.Instance.CreateParam("_id_ejemplar", DbType.Int32, idObjeto, ParameterDirection.Input);
            return DBManager.Instance.EjecutarProcedimiento("ELIMINAR_EJEMPLAR", parametros);
        }

        public int insertar(Ejemplar ejemplar)
        {
            DbParameter[] parametros = new DbParameter[4];
            parametros[0] = DBManager.Instance.CreateParam("_id_ejemplar", DbType.Int32, null, ParameterDirection.Output);
            parametros[1] = DBManager.Instance.CreateParam("_id_material", DbType.Int32, ejemplar.Material.IdMaterial, ParameterDirection.Input);
            parametros[2] = DBManager.Instance.CreateParam("_ubicacion", DbType.String, ejemplar.Ubicacion, ParameterDirection.Input);
            parametros[3] = DBManager.Instance.CreateParam("_id_biblioteca", DbType.Int32, ejemplar.Biblioteca.IdBiblioteca, ParameterDirection.Input);
            DBManager.Instance.EjecutarProcedimiento("INSERTAR_EJEMPLAR", parametros);
            ejemplar.IdEjemplar = Convert.ToInt32(parametros[0].Value);
            return ejemplar.IdEjemplar;
        }

        public BindingList<Ejemplar> listarTodos()
        {
            BindingList<Ejemplar> ejemplares = null;
            lector = DBManager.Instance.EjecutarProcedimientoLectura("LISTAR_EJEMPLARES_TODOS", null);
            while (lector.Read())
            {
                if (ejemplares == null) ejemplares = new BindingList<Ejemplar>();
                Ejemplar ejemplar = new Ejemplar();
                if (!lector.IsDBNull(lector.GetOrdinal("id_ejemplar"))) ejemplar.IdEjemplar = lector.GetInt32(lector.GetOrdinal("id_ejemplar"));// Se coloca el identificador del select
                if (!lector.IsDBNull(lector.GetOrdinal("id_material"))) ejemplar.Material.IdMaterial = lector.GetInt32(lector.GetOrdinal("id_material"));
                if (!lector.IsDBNull(lector.GetOrdinal("estado"))) ejemplar.Estado = (EstadoEjemplar)Enum.Parse(typeof(EstadoEjemplar), lector.GetString(lector.GetOrdinal("estado")));
                if (!lector.IsDBNull(lector.GetOrdinal("ubicacion"))) ejemplar.Ubicacion = lector.GetString(lector.GetOrdinal("ubicacion"));
                if (!lector.IsDBNull(lector.GetOrdinal("activo"))) ejemplar.Activo = lector.GetBoolean(lector.GetOrdinal("activo"));
                if (!lector.IsDBNull(lector.GetOrdinal("id_biblioteca"))) ejemplar.Biblioteca.IdBiblioteca = lector.GetInt32(lector.GetOrdinal("id_biblioteca"));
                ejemplares.Add(ejemplar);
            }
            DBManager.Instance.CerrarConexion();
            return ejemplares;
        }

        public int modificar(Ejemplar ejemplar)
        {
            DbParameter[] parametros = new DbParameter[4];
            parametros[0] = DBManager.Instance.CreateParam("_id_ejemplar", DbType.Int32, ejemplar.IdEjemplar, ParameterDirection.Output);
            parametros[1] = DBManager.Instance.CreateParam("_id_material", DbType.Int32, ejemplar.Material.IdMaterial, ParameterDirection.Input);
            parametros[2] = DBManager.Instance.CreateParam("_ubicacion", DbType.String, ejemplar.Ubicacion, ParameterDirection.Input);
            parametros[3] = DBManager.Instance.CreateParam("_id_biblioteca", DbType.Int32, ejemplar.Biblioteca.IdBiblioteca, ParameterDirection.Input);
            return DBManager.Instance.EjecutarProcedimiento("MODIFICAR_EJEMPLAR", parametros);
        }

        public Ejemplar obtenerPorId(int idObjeto)
        {
            Ejemplar ejemplar = null;
            DbParameter[] parametros = new DbParameter[1];
            parametros[0] = DBManager.Instance.CreateParam("_id_ejemplar", DbType.Int32, idObjeto, ParameterDirection.Input);
            lector = DBManager.Instance.EjecutarProcedimientoLectura("OBTENER_EJEMPLAR_X_ID", parametros);
            if (lector.Read())
            {
                if (ejemplar == null) ejemplar = new Ejemplar();
                if (!lector.IsDBNull(lector.GetOrdinal("id_ejemplar"))) ejemplar.IdEjemplar = lector.GetInt32(lector.GetOrdinal("id_ejemplar"));// Se coloca el identificador del select
                if (!lector.IsDBNull(lector.GetOrdinal("id_material"))) ejemplar.Material.IdMaterial = lector.GetInt32(lector.GetOrdinal("id_material"));
                if (!lector.IsDBNull(lector.GetOrdinal("estado"))) ejemplar.Estado = (EstadoEjemplar)Enum.Parse(typeof(EstadoEjemplar), lector.GetString(lector.GetOrdinal("estado")));
                if (!lector.IsDBNull(lector.GetOrdinal("ubicacion"))) ejemplar.Ubicacion = lector.GetString(lector.GetOrdinal("ubicacion"));
                if (!lector.IsDBNull(lector.GetOrdinal("activo"))) ejemplar.Activo = lector.GetBoolean(lector.GetOrdinal("activo"));
                if (!lector.IsDBNull(lector.GetOrdinal("id_biblioteca"))) ejemplar.Biblioteca.IdBiblioteca = lector.GetInt32(lector.GetOrdinal("id_biblioteca"));
            }
            DBManager.Instance.CerrarConexion();
            return ejemplar;
        }

        public BindingList<Ejemplar> listar_disponibles_por_material(int _id_material)
        {
            BindingList<Ejemplar> ejemplares = null;
            DbParameter[] parametros = new DbParameter[1];
            parametros[0] = DBManager.Instance.CreateParam("_id_material", DbType.Int32, _id_material, ParameterDirection.Input);

            lector = DBManager.Instance.EjecutarProcedimientoLectura("LISTAR_EJEMPLARES_DISPONIBLES_POR_MATERIAL", parametros);
            while (lector.Read())
            {
                if (ejemplares == null) ejemplares = new BindingList<Ejemplar>();
                Ejemplar ejemplar = new Ejemplar();
                if (!lector.IsDBNull(lector.GetOrdinal("id_ejemplar"))) ejemplar.IdEjemplar = lector.GetInt32(lector.GetOrdinal("id_ejemplar"));// Se coloca el identificador del select
               MaterialBibliografico material = new MaterialBibliografico();
                if (!lector.IsDBNull(lector.GetOrdinal("id_material"))) material.IdMaterial = lector.GetInt32(lector.GetOrdinal("id_material"));
                ejemplar.Material = material;
                if (!lector.IsDBNull(lector.GetOrdinal("estado"))) ejemplar.Estado = (EstadoEjemplar)Enum.Parse(typeof(EstadoEjemplar), lector.GetString(lector.GetOrdinal("estado")));
                if (!lector.IsDBNull(lector.GetOrdinal("ubicacion"))) ejemplar.Ubicacion = lector.GetString(lector.GetOrdinal("ubicacion"));
                Biblioteca biblioteca = new Biblioteca();
                if (!lector.IsDBNull(lector.GetOrdinal("id_biblioteca"))) biblioteca.IdBiblioteca = lector.GetInt32(lector.GetOrdinal("id_biblioteca"));
                ejemplar.Biblioteca = biblioteca;
                ejemplares.Add(ejemplar);
            }
            DBManager.Instance.CerrarConexion();



            return ejemplares;
        }
    }
}

File: BackEnd (Java)/BibliotecaDA/src/main/java/biblioteca/gestionMaterial/mysql/EjemplarImpl.java

package biblioteca.gestionMaterial.mysql;

import biblioteca.config.DBManager;
import biblioteca.gestionMaterial.dao.EjemplarDAO;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import pe.edu.pucp.utilsarmy.gestion_de_material.model.Biblioteca;
import pe.edu.pucp.utilsarmy.gestion_de_material.model.Ejemplar;
import pe.edu.pucp.utilsarmy.gestion_de_material.model.EstadoEjemplar;

public class EjemplarImpl implements EjemplarDAO {
private ResultSet rs;
    @Override
    public int insertar(Ejemplar objeto) {
        Map<Integer, Object> parametrosSalida = new HashMap<>();
        Map<Integer, Object> parametrosEntrada = new HashMap<>();
        parametrosSalida.put(1, Types.INTEGER);
        parametrosEntrada.put(2, objeto.getId_material());
        parametrosEntrada.put(3, objeto.getUbicacion());
        parametrosEntrada.put(4, objeto.getBlibioteca().getIdBiblioteca());
        DBManager.getInstance().ejecutarProcedimiento("INSERTAR_Ejemplar", parametrosEntrada, parametrosSalida);
        objeto.setIdEjemplar((int) parametrosSalida.get(1));
        System.out.println("Se ha realizado el registro del ejemplar");
        return objeto.getIdEjemplar();
    }

    @Override
    public int modificar(Ejemplar objeto) {
        Map<Integer, Object> parametrosEntrada = new HashMap<>();
        parametrosEntrada.put(1, objeto.getIdEjemplar());
        parametrosEntrada.put(2, objeto.getEstado().name());
        parametrosEntrada.put(3, objeto.getUbicacion());
        parametrosEntrada.put(4, objeto.isActivo());
        parametrosEntrada.put(5, objeto.getBlibioteca().getIdBiblioteca());
        int resultado = DBManager.getInstance().ejecutarProcedimiento("MODIFICAR_EJEMPLAR", parametrosEntrada, null);
        System.out.println("Se ha realizado la modificacion del ejemplar");
        return resultado;
    }

    @Override
    public int eliminar(int idObjeto) {
        Map<Integer, Object> parametrosEntrada = new HashMap<>();
        parametrosEntrada.put(1, idObjeto);
        int resultado = DBManager.getInstance().ejecutarProcedimiento("ELIMINAR_EJEMPLAR", parametrosEntrada, null);
        System.out.println("Se ha realizado la eliminacion del ejemplar");
        return resultado;
    }

    @Override
    public Ejemplar obtenerPorId(int idObjeto) {
        Ejemplar ejemplar = null;
        Map<Integer, Object> parametrosEntrada = new HashMap<>();
        parametrosEntrada.put(1, idObjeto);

        ResultSet rs = DBManager.getInstance().ejecutarProcedimientoLectura("OBTENER_EJEMPLAR_X_ID", parametrosEntrada);
        System.out.println("Lectura de ejemplar...");
        try {
            if (rs.next()) {
                ejemplar = new Ejemplar();
                ejemplar.setIdEjemplar(rs.getInt("id_ejemplar"));
                ejemplar.setId_material(rs.getInt("id_material"));
                ejemplar.setUbicacion(rs.getString("ubicacion"));
                ejemplar.setActivo(rs.getBoolean("activo"));
                Biblioteca b=new Biblioteca();
                b.setIdBiblioteca(rs.getInt("id_biblioteca"));
                ejemplar.setBlibioteca(b);
                
                String estadoStr = rs.getString("estado");
                ejemplar.setEstado(EstadoEjemplar.valueOf(estadoStr.toUpperCase()));
            }
        } catch (SQLException ex) {
            System.out.println("ERROR: " + ex.getMessage());
        } finally {
            DBManager.getInstance().cerrarConexion();
        }
        return ejemplar;
    }

    @Override
    public ArrayList<Ejemplar> listarTodos() {
        ArrayList<Ejemplar> ejemplares = null;
        rs = DBManager.getInstance().ejecutarProcedimientoLectura("LISTAR_EJEMPLARES_TODOS", null);
        System.out.println("Lectura de ejemplares...");
        try {
            while (rs.next()) {
                if (ejemplares == null) ejemplares = new ArrayList<>();
                Ejemplar e = new Ejemplar();

                e = new Ejemplar();
                e.setIdEjemplar(rs.getInt("id_ejemplar"));
                e.setId_material(rs.getInt("id_material"));
                e.setUbicacion(rs.getString("ubicacion"));
                e.setActivo(rs.getBoolean("activo"));
                Biblioteca b=new Biblioteca();
                b.setIdBiblioteca(rs.getInt("id_biblioteca"));
                e.setBlibioteca(b);
                
                String estadoStr = rs.getString("estado");
                e.setEstado(EstadoEjemplar.valueOf(estadoStr.toUpperCase()));
                ejemplares.add(e);
            }
        } catch (SQLException ex) {
            System.out.println("ERROR: " + ex.getMessage());
        } finally {
            DBManager.getInstance().cerrarConexion();
        }
        return ejemplares;
    }

    @Override
    public ArrayList<Ejemplar> listar_disponibles_por_material(int _id_material) {
        ArrayList<Ejemplar> ejemplares = null;
        Map<Integer, Object> parametrosEntrada = new HashMap<>();
        parametrosEntrada.put(1, _id_material);
        rs = DBManager.getInstance().ejecutarProcedimientoLectura("LISTAR_EJEMPLARES_DISPONIBLES_POR_MATERIAL", parametrosEntrada);
        System.out.println("Lectura de ejemplares...");
        try {
            while (rs.next()) {
                if (ejemplares == null) ejemplares = new ArrayList<>();
                Ejemplar e = new Ejemplar();

                e = new Ejemplar();
                e.setIdEjemplar(rs.getInt("id_ejemplar"));
                e.setId_material(rs.getInt("id_material"));
                e.setUbicacion(rs.getString("ubicacion"));
                Biblioteca b=new Biblioteca();
                b.setIdBiblioteca(rs.getInt("id_biblioteca"));
                e.setBlibioteca(b);
                
                String estadoStr = rs.getString("estado");
                e.setEstado(EstadoEjemplar.valueOf(estadoStr.toUpperCase()));
                ejemplares.add(e);
            }
        } catch (SQLException ex) {
            System.out.println("ERROR: " + ex.getMessage());
        } finally {
            DBManager.getInstance().cerrarConexion();
        }
        return ejemplares;
    }
    

}
